modify : session 기능 수정, currentMember API 추가

corea1836 opened the pull request:
## Description
- SessionLoginService와 SessionUtil 책임 분리
     - LoginService :  로그인에 대한 비지니스 로직 책임
     - SessionUtil : 세션에 대한 로직 책임

-  모든 U/D와 일부 C API에 로그인 유저가 파라미터로 동작하도록 수정
    - Review, HospitalAdmin Create API가 현재 새션 유저를 파라미터로 받도록 수정
    - HospitalAdmin U, D API가 현재 새션 유저 == 병원 관리자 인지 확인하는 기능 추가 
    - 나머지 U, D API에 현재 세션 유저가 작성한 내용인지 확인하는 기능 추가

- memberUniqueId(Long) -> loginId(String) 으로 파라미터 수정

corea1836 committed:
feat: SessionUtil 생성

- SessionLoginService와 SessionUtil 책임 분리
  - LoginService :  로그인에 대한 비지니스 로직 책임
  - SessionUtil : 세션에 대한 로직 책임

corea1836 committed:
modify : 모든 C/U/D에 로그인 유저 파라미터 추가

- 모든 U/D와 일부 C API에 로그인 유저가 파라미터로 동작하도록 수정

corea1836 commented on src/main/java/flab/docdoc/common/util/SessionUtil.java:
```diff
@@ -0,0 +1,36 @@
+
+@Component
+@RequiredArgsConstructor
+public class SessionUtil {
```
세션에 대한 책임을 분리 했습니다.

AS IS
SessionLoginService : 세션 기반으로 로그인, 로그아웃 상태를 변경, 확인

TO BE
SessionUtil : 세션과 관련된 로직을 수행
LoginService : 로그인과 관련된 비지니스 로직을 수행

corea1836 commented on src/main/java/flab/docdoc/common/util/SessionUtil.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    private static HttpSession getSession() {
+        return ((ServletRequestAttributes) RequestContextHolder.currentRequestAttributes()).getRequest().getSession();
```
Util로 만들었기 때문에 전역에서 사용할 수 있도록 static 메서드로 만들었습니다.
세션에 관한 로직을 수행하려면 HttpSession을 사용해야하는데 아래 세 가지 방법 중 고민했습니다.
- DI 받는 방법 -> X : DI를 받으려면 non-static을 사용해야하는데, Util은 static 메서드이기 때문에 로드 타임이 다릅니다.
- 파라미터로 받는 방법 -> X : 파라미터로 request에서 HttpSession을 추출해 파라미터로 넘기는 방법도 있지만 모든 로직에 추가해야한다는 점이 번거롭습니다.
- RequestContextHolder -> O : 스프링에 의존적이긴 하지만 세션에 관한 역할을 분리할 수 있다는 점이 가장 적합하다 생각해 사용했습니다. 내부적으로 ThreadLocal을 사용하고 있기 때문에 Thread safe 합니다.

f-lab-kyle commented on src/main/java/flab/docdoc/common/util/SessionUtil.java:
```diff
@@ -0,0 +1,36 @@
+
+    private static final String LOGIN_MEMBER_ID = "memberId";
+
+    public static String getLoginMemberId() {
+        return LOGIN_MEMBER_ID;
+    }
```
이건 외부 노출이 필요 없지 않을까 싶네요 + 메소드로 표현 안해도 정적 변수 네이밍으로 충분할것 같아요.

corea1836 replied:
🙆‍♂️

f-lab-kyle commented on src/main/java/flab/docdoc/hospital/controller/HospitalAdminController.java:
```diff
@@ -20,13 +22,16 @@ public class HospitalAdminController {
     @PutMapping("/add")
     public ResponseEntity<HttpStatus> updateAdmin(@RequestBody @Valid UpdateHospitalAdminRequest request) {
-        hospitalAdminService.addHospitalAdmin(request);
+        String loginId = (String) getCurrentMember().orElseThrow(() -> {throw new IllegalArgumentException("로그인 상태가 아닙니다.");});
+
+        hospitalAdminService.addHospitalAdmin(request, loginId);
```
아래도 그렇고 위에서 만든 AuthAspect 를 사용하면 안되려나요?

corea1836 replied:
🙆‍♂️

f-lab-kyle commented on src/main/java/flab/docdoc/hospital/service/HospitalAdminService.java:
```diff
@@ -4,7 +4,7 @@
 public interface HospitalAdminService {
 
-    public void addHospitalAdmin(UpdateHospitalAdminRequest request);
+    public void addHospitalAdmin(UpdateHospitalAdminRequest request, final String loginId);
```
request DTO에 담을순 없으려나요

corea1836 replied:
🙆‍♂️

f-lab-kyle commented on src/main/java/flab/docdoc/hospital/service/HospitalServiceImpl.java:
```diff
@@ -47,8 +47,15 @@ public void save(AddHospitalRequest request) {
+        });
+
+
+        if(!existHospital.getAdminId().equals(loginId)) {
```
그럴일은 적지만, 이런 String 비교에는 npe를 고려하여 StringUtils.equals(A,B) 를 사용하곤 합니다.
만약, existHospital.getAdminId() 이 null일수도 있으니까요. (String이라는 참조형 변수의 위험성)

corea1836 replied:
도메인 별 테스트 브랜치에서 수정 하겠습니다.

f-lab-kyle commented on src/main/java/flab/docdoc/member/service/SessionLoginService.java:
```diff
@@ -7,14 +7,13 @@
 import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Service;
 
+import static flab.docdoc.common.util.SessionUtil.*;
```
가급적 * 은 사용 안하는 걸로~ 
사용하지 않는 녀석들도 import하기도 하고 의존성 체크가 어려워요

corea1836 replied:
🙆‍♂️

corea1836 committed:
Merge branch 'main' into modify_get_session_member

corea1836 committed:
modify : CheckLogin, CheckWriter Aspect 추가

- CheckLogin Aspect 추가 : 로그인 된 멤버인지 세션을 통해 확인
- CheckWriter Aspect 추가 : 요청한 Request의 글쓴이가 로그인 멤버인지 확 인

- WriterMatch I/F 추가 : CheckWriter Aspect가 검사할 때 쓰는 메서드를 포함
